### ISSUE TITLE
max_active_size can cause deadlock when deferring items in a pipeline

### ISSUE THREAD

**Issue by djay** [scrapy/scrapy#6206]
### Description

When you hit the SCRAPER_SLOT_MAX_ACTIVE_SIZE requests get dropped silently with no warning.
If you are deferring items in a pipeline then your crawler will never finish those requests if they whats needed to finish the original deferred. So it's a deadlock with no explanation which took a long time to work out why.
This is made worse by the prioritydownloader as this preferences less active sites so the bottleneck site causing the deadlock never gets the requests as they are the ones dropped.

[[Description of the issue]](https://github.com/scrapy/scrapy/pull/3551)

### Steps to Reproduce

1. Run a crawl with a pipeline that will defer items. This will mean you hit max_active_size fast.
2. Notice no warning or error
3. [and so on...]

**Expected behavior:** [What you expect to happen]

An error when you hit max_active_size. Stats addon to show number of active requests.

**Actual behavior:** [What actually happens]

Nothing shown so hard to debug

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

2.11.0, 2.7.0

Please paste here the output of executing `scrapy version --verbose` in the command line.

```
Scrapy       : 2.11.0
lxml         : 4.9.1.0
libxml2      : 2.9.14
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 2.0.1
Twisted      : 22.8.0
Python       : 3.9.18 (main, Oct  3 2023, 08:56:37) - [Clang 12.0.0 (clang-1200.0.32.2)]
pyOpenSSL    : 22.1.0 (OpenSSL 3.0.5 5 Jul 2022)
cryptography : 38.0.1
Platform     : macOS-10.16-x86_64-i386-64bit
```

### Additional context

Any additional information, configuration, data or output from commands that might be necessary to reproduce or understand the issue. Please try not to include screenshots of code or the command line, paste the contents as text instead. You can use [GitHub Flavored Markdown](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) to make the text look better.

**Comment by Gallaecio** [comment]
I guess we could add a warning the first time the limit is reached, in case uses want to consider increasing it (to improve performance if they can spare the memory). But we should also allow disabling that warning.

**Comment by djay** [comment]
@Gallaecio I'd suggest in the stats addon to show an indication only when over max_active_size. Currently there is nothing to show there is even anything in the queue so it wasn't obvious the problem wasn't items being dropped via errors or something vs just the queue being paused. or alternatively something was slowing requests way down.

**Comment by GeorgeA92** [comment]
> When you hit the SCRAPER_SLOT_MAX_ACTIVE_SIZE requests stop being processed silently with no warning.
> 
> **Expected behavior:** 
> An error when you hit max_active_size. Stats addon to show number of active requests.

this is.. core feature by design - not a bug. If application accumulated some amount of not processed responses - it stops to send new requests (to prevent RAM overhead).
If application really got stuck because of `SCRAPER_SLOT_MAX_ACTIVE_SIZE` - it most likely mean that it got stuck because of CPU bottleneck (response processing rate - slower than request sending rate) and in this case it's most likely duplicate of https://github.com/scrapy/scrapy/issues/5302


> If you are deferring items in a pipeline that depend on other requests finishing before completing then you can get into a deadlock with little explanation why.

It depends on the way how exactly new requests called inside pipeline (If I correctly interpret meaning of "deferring items in a pipeline that depend on other requests..")
For example image requests from Scrapy ImagePipeline - doesn't affect scraper.slot.active_size. 
Scraper slot is expected to affect responses that aimed to be processed by spider callback methods. Responses from other requests(called from pipeline) is not expected to appear to `scraper` (and to increase it's active_size value).
Also requests from from ImagePipeline - do not appear on scheduler which means that `DownloaderAwarePriorityQueue` didn't applied to them.


Can You please provide.. more detailed info at least dump stats log output (ending log entry starting with `[scrapy.statscollectors] INFO: Dumping Scrapy stats:`

> as this preferences **less active sites so the bottleneck** site causing...

Is this a broadcrawl case?

**Comment by GeorgeA92** [comment]
https://github.com/scrapy/scrapy/blob/d5233bb57f35c54b0c03981dc59c7328ca44cb9a/scrapy/core/engine.py#L197-L205
This code fragment define conditions when application send (or not send) new request:
1.  `not self.running or bool(self.slot.closing)` if closespider signal triggered (not this case)
2.  `self.downloader.needs_backout()` if Downloader queue reached size of `CONCURRENT_REQUESTS`
3. `self.scraper.slot.needs_backout()` if application accumulated some not processed responses - `SCRAPER_SLOT_MAX_ACTIVE_SIZE` check.

@djay 
How did You conclude that issue is because of `SCRAPER_SLOT_MAX_ACTIVE_SIZE` (not busy downloader)?
If it is really broadcrawl - I'd bet on version with downloader bottleneck by timeout exceptions (not scraper slot active size) mentioned on https://docs.scrapy.org/en/latest/topics/broad-crawls.html?highlight=broad-crawls#reduce-download-timeout

But it will be clear only after at least review of `[scrapy.statscollectors] INFO: Dumping Scrapy stats` log entry I requested.

**Comment by djay** [comment]
> Can You please provide.. more detailed info at least dump stats log output 
```
...
2024-01-17 00:16:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:17:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:18:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:19:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:20:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:21:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:22:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:23:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
2024-01-17 00:24:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
```

> For example image requests from Scrapy ImagePipeline - doesn't affect scraper.slot.active_size

I will look into how that is done because this is the issue I'm having. To collect more information to include in an item I have a pipeline that generates a lot of requests that do increase active_size and this creates teh deadlock. Increasing SCRAPER_SLOT_MAX_ACTIVE_SIZE prevents the problem for longer.

**Comment by GeorgeA92** [comment]
> 2024-01-17 00:16:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
> 2024-01-17 00:17:58 [scrapy.extensions.logstats] INFO: Crawled 2678 pages (at 0 pages/min), scraped 1636 items (at 0 items/min)
> 2024-01-17 00:18:58..

As I mentioned earlier It still can be caused by timeouts. Request that returned timeout Exception (no Response returned) - doesn't count as crawled. During that period that from provided log entries looks like idle - application may send hundreds or even thousands of requests (with no responses). That why I've asked for `[scrapy.statscollectors] INFO: Dumping Scrapy stats` log entry - not log entries from ` [scrapy.extensions.logstats]`
Alternatively You can try to apply `PeriodicLog` extension that log required stats periodically(and post it's output here) mentioned on. https://docs.scrapy.org/en/latest/topics/extensions.html#periodic-log-extension.

> To collect more information to include in an item I have a pipeline that generates a lot of requests **that do increase active_size** and this creates teh deadlock

We still don't know this for sure. Calling new requests from pipeline is not very good practice. And as I mentioned earlier requests originated from pipeline may not reach scheduler, scraper(do not directly affect it's slot active size). It'd be good to share pipeline code here too.

**Comment by djay** [comment]
> We still don't know this for sure. 
99% sure. If I increase SCRAPER_SLOT_MAX_ACTIVE_SIZE it solves the problem. I also used the debugger and the AUTOTHROTTLE_DEBUG=True to show that no requests are happening.

> Calling new requests from pipeline is not very good practice.
That maybe but my main point is that there is no warning about this. It's not in the documentation. It's not in the log messages. And really if you are stopping the the queue getting processed by hitting some limit I think it's reasonable to assume there should be some log message for this right?

The documention only says a pipeline can return a Deferred. And that seems like a reasonable way to combine information from two different crawls into a single item. I don't actually spawn a request in the pipeline. I do it in another middleware. That then sets a special value in the item which includes a deferred for when the history crawl has been completed and a callback to set values on the item to replace the special values with the real values from the history.
This seemed what returning a Deferred was designed for. What is unexpected is that doing so keeps the original request as still active despite request or response not being involved in the pipeline. I would not have expected that and that is partly why it was so confusing that using Deferred in the pipeline caused a problem.

``` 
class DeferHistoryPipeline:
    """ Add in history once history async is complete """
    def process_item(self, item, spider):
        # return item
        i = ItemAdapter(item)
        d = None
        callbacks = []
        for value in i.values():
            if type(value) == tuple and isinstance(value[0], Deferred):
                d, callback = value
                callbacks.append(callback)
        if not callbacks:
            return item

        def do_callbacks(history):
            for needs_item in callbacks:
                needs_item(i, history)
            return item
        return d.addBoth(do_callbacks)
```

Note I also tried to solve this by setting my history requests to be higher priority but it doesn't seem to be respected so that didn't work. If priority worked then the history requests would be completed before new items are generated and then deferred waiting for even more history requests, so the active_size doesn't increase as fast. 
Or if the requests were forgotten by the pipeline step so didn't increase the active_size. 
Either of these would be nice. But really all I'm asking for is a warning so others don't waste as much time as I did  trying to debug something which is apparently not catered for.

**Comment by djay** [comment]
> Also requests from from ImagePipeline - do not appear on scheduler which means that DownloaderAwarePriorityQueue didn't applied to them.

it seems like the comment below isnt the case according to documentation below.

It also seems like it is standard practice to use Deferred and spawn requests that depend on it from the pipeline - see https://docs.scrapy.org/en/latest/topics/media-pipeline.html 

> When the item reaches the FilesPipeline, the URLs in the file_urls field are scheduled for download using the standard Scrapy scheduler and downloader (which means the scheduler and downloader middlewares are reused), but with a higher priority, processing them before other pages are scraped. The item remains “locked” at that particular pipeline stage until the files have finish downloading (or fail for some reason).

What this isn't telling you is that if the files are from a single domain  and other urls aren't, then priority won't really help since the if you have any kind of scheduling that limits requests per domain since then other requests from other domains will skip ahead of hte priority requests, create more file requests and you can end up a deadlock. And have no idea why.

**Comment by djay** [comment]
> Also requests from from ImagePipeline - do not appear on scheduler which means that DownloaderAwarePriorityQueue didn't applied to them.

ok. I think I see that self.cawler.engine.download instead of engine.crawl is the special sauce that is needed to prevent this

https://github.com/scrapy/scrapy/blob/master/scrapy/pipelines/media.py#L144

https://github.com/scrapy/scrapy/blob/d5233bb57f35c54b0c03981dc59c7328ca44cb9a/scrapy/core/engine.py#L312

**Comment by djay** [comment]
Crawler.download doesn't seem to help. Even if it avoids adding to the active size the active size still grows and eventually blocks all requests thus preventing it shrinking. 

Why is a deferred item in a pipeline still considered part of active requests? That is the problem here. All that's left in the pipeline is the item. The request abd response aren't accessible. They are done.  They can't be reversed.

**Comment by djay** [comment]
@Gallaecio I'm not 100% sure but I don't think this test shows that the active size didn't increase while images were being downloaded, (ie before the deferred is released). It only shows the active size was the same after the Deferred was released? You can see a-light-in-the-attic_1000 resulted in 4 images and it only reported the active size once all had be downloaded.
But if it is the case then somehow I must have not used engine.download in some part of my code that I missed and that might have prevented the deadlock.
In any case I've rewritten my code to not use a ItemPipeline and now don't get deadlocks. I still think it should be more obvious somehow to see max active size has been exceeded and show what it is waiting on to make it more obvious a deadlock is happening?

**Comment by GeorgeA92** [comment]
@djay 

> I'm not 100% sure but I don't think this test shows that the active size didn't increase while images were being downloaded, (ie before the deferred is released). It only shows the active size was the same after the Deferred was released?

Key point of this test is that original html response that already finished it's parse method - still counted as active.
Users may expect that after completion of response parse method - slot active size - should be to 0 (right after ` [__main__] INFO: parse method finished !!!!` log entry) which didn't happend here.
On a real case - multiple html responses that already parsed(and not needed) can cause deadlock if their total size exceed scraper slot max active size

```
2024-06-20 14:48:14 [__main__] INFO: scraper active_size is: 0
2024-06-20 14:48:15 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/> (referer: None)
2024-06-20 14:48:15 [__main__] INFO: scraper active_size is: 51294
2024-06-20 14:48:15 [__main__] INFO: parse method finished !!!!
2024-06-20 14:48:16 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/media/cache/2c/da/2cdad67c44b002e7ead0cc35693c0e8b.jpg> (referer: None)
2024-06-20 14:48:16 [scrapy.pipelines.files] DEBUG: File (downloaded): Downloaded file from <GET http://books.toscrape.com/media/cache/2c/da/2cdad67c44b002e7ead0cc35693c0e8b.jpg> referred in <None>
2024-06-20 14:48:16 [scrapy.core.scraper] DEBUG: Scraped from <200 http://books.toscrape.com/>
{'url': 'catalogue/a-light-in-the-attic_1000/index.html', 'image_urls': ['http://books.toscrape.com/media/cache/2c/da/2cdad67c44b002e7ead0cc35693c0e8b.jpg'], 'images': [{'url': 'http://books.toscrape.com/media/cache/2c/da/2cdad67c44b002e7ead0cc35693c0e8b.jpg', 'path': 'full/22823e0c52cac8fe35a2c7d355753481bfda423b.jpg', 'checksum': '43717a947a533150c3d152a07dd9e883', 'status': 'downloaded'}]}
2024-06-20 14:48:17 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/media/cache/26/0c/260c6ae16bce31c8f8c95daddd9f4a1c.jpg> (referer: None)
2024-06-20 14:48:17 [scrapy.pipelines.files] DEBUG: File (downloaded): Downloaded file from <GET http://books.toscrape.com/media/cache/26/0c/260c6ae16bce31c8f8c95daddd9f4a1c.jpg> referred in <None>
2024-06-20 14:48:17 [scrapy.core.scraper] DEBUG: Scraped from <200 http://books.toscrape.com/>
{'url': 'catalogue/tipping-the-velvet_999/index.html', 'image_urls': ['http://books.toscrape.com/media/cache/26/0c/260c6ae16bce31c8f8c95daddd9f4a1c.jpg'], 'images': [{'url': 'http://books.toscrape.com/media/cache/26/0c/260c6ae16bce31c8f8c95daddd9f4a1c.jpg', 'path': 'full/d3d729396ae53426cee85e8adc77f9d288c9b7cc.jpg', 'checksum': '6f34fd0b23a28c389ec1a697d4cd539f', 'status': 'downloaded'}]}
2024-06-20 14:48:18 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/media/cache/3e/ef/3eef99c9d9adef34639f510662022830.jpg> (referer: None)
2024-06-20 14:48:18 [scrapy.pipelines.files] DEBUG: File (downloaded): Downloaded file from <GET http://books.toscrape.com/media/cache/3e/ef/3eef99c9d9adef34639f510662022830.jpg> referred in <None>
2024-06-20 14:48:18 [scrapy.core.scraper] DEBUG: Scraped from <200 http://books.toscrape.com/>
{'url': 'catalogue/soumission_998/index.html', 'image_urls': ['http://books.toscrape.com/media/cache/3e/ef/3eef99c9d9adef34639f510662022830.jpg'], 'images': [{'url': 'http://books.toscrape.com/media/cache/3e/ef/3eef99c9d9adef34639f510662022830.jpg', 'path': 'full/f0ba24b1ce745331dd260f9e922248c1a6d5c866.jpg', 'checksum': '747b8bb69aad837541306a8732ae3370', 'status': 'downloaded'}]}
2024-06-20 14:48:19 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/media/cache/32/51/3251cf3a3412f53f339e42cac2134093.jpg> (referer: None)
2024-06-20 14:48:19 [scrapy.pipelines.files] DEBUG: File (downloaded): Downloaded file from <GET http://books.toscrape.com/media/cache/32/51/3251cf3a3412f53f339e42cac2134093.jpg> referred in <None>
2024-06-20 14:48:19 [scrapy.core.scraper] DEBUG: Scraped from <200 http://books.toscrape.com/>
{'url': 'catalogue/sharp-objects_997/index.html', 'image_urls': ['http://books.toscrape.com/media/cache/32/51/3251cf3a3412f53f339e42cac2134093.jpg'], 'images': [{'url': 'http://books.toscrape.com/media/cache/32/51/3251cf3a3412f53f339e42cac2134093.jpg', 'path': 'full/c19b9ddd65b792cb227f1284fc7fefcf0f9062ef.jpg', 'checksum': 'e578fcda7210a0770de667e187b123cf', 'status': 'downloaded'}]}
2024-06-20 14:48:21 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://books.toscrape.com/media/cache/be/a5/bea5697f2534a2f86a3ef27b5a8c12a6.jpg> (referer: None)
2024-06-20 14:48:21 [scrapy.pipelines.files] DEBUG: File (downloaded): Downloaded file from <GET http://books.toscrape.com/media/cache/be/a5/bea5697f2534a2f86a3ef27b5a8c12a6.jpg> referred in <None>
2024-06-20 14:48:21 [__main__] INFO: scraper active_size is: 51294
```

**Comment by Gallaecio** [comment]
> On a real case - multiple html responses that already parsed(and not needed) can cause deadlock if their total size exceed scraper slot max active size

I still don’t understand in which scenario there is a deadlock, if by deadlock we understand that the spider will stop processing responses forever.

What would prevent the active size from going down? In what scenario?

**Comment by djay** [comment]
@Gallaecio you have an itempipeline that returns Defered objects in it's process_item() method.
It also generates a bunch more requests. 
The Deferred is not marked completed until those requests are processed. 
Unless there is a way to ensure the requests coming from the pipeline are done before other requests then eventually you get a situation where you've reached the max active is met with all those deferred. This then blocks any more requests being processed and you get a deadlock.

**Comment by Gallaecio** [comment]
If those requests are executed with `engine.download`, as opposed to being scheduled (which you would not want to do in this scenario), they are not affected by this setting. This setting only prevents requests from start_requests and from the scheduler from being processed, not requests sent with engine.download.

**Comment by djay** [comment]
So I've been told. But I was using engine.downloaf and it was still going to into deadlock. Nothing being processed. The requests processed stats remains exactly the same every minute for hours. 

I've now resolved it by using spidermiddeware instead and using callbacks in such a way that the last extra requested generated is the one that returns the item. This means there are no active requests hanging around waiting. 
This seems to work well. I still had to use priority to ensure the sub requests got done earlier so items came out quicker but that wasn't to resolve a deadlock. 

The original point of this issue still stands however. If max active is hit and no more requests are being processed there really should be some warning to let you know. That would help someone realise they have created a deadlock.

**Comment by Gallaecio** [comment]
If you believe that the scenario with `engine.download()` creates a deadlock, please provide a minimal reproducible example of it, so that we can verify the issue, and discard it being caused by something else.

>  I still had to use priority to ensure the sub requests got done earlier so items came out quicker but that wasn't to resolve a deadlock.

Using `engine.download` for this would be better, as it bypasses the scheduler, no need to set a priority at all to get them to run as soon as possible.

> If max active is hit and no more requests are being processed there really should be some warning to let you know.

https://github.com/scrapy/scrapy/pull/6430 should address this.
